legend-engine-executionPlan-execution/src/main/java/org/finos/legend/engine/plan/execution/concurrent/ConcurrentExecutionNodeExecutorPool.java

```diff
@@ -0,0 +1,158 @@
+//  Copyright 2023 Goldman Sachs
+//
+//  Licensed under the Apache License, Version 2.0 (the "License");
+//  you may not use this file except in compliance with the License.
+//  You may obtain a copy of the License at
+//
+//       http://www.apache.org/licenses/LICENSE-2.0
+//
+//  Unless required by applicable law or agreed to in writing, software
+//  distributed under the License is distributed on an "AS IS" BASIS,
+//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+
+package org.finos.legend.engine.plan.execution.concurrent;
+
+import io.opentracing.Scope;
+import io.opentracing.util.GlobalTracer;
+import org.eclipse.collections.api.list.MutableList;
+import org.eclipse.collections.api.tuple.Pair;
+import org.eclipse.collections.impl.factory.Lists;
+import org.eclipse.collections.impl.list.mutable.FastList;
+import org.eclipse.collections.impl.tuple.Tuples;
+import org.eclipse.collections.impl.utility.ListIterate;
+import org.finos.legend.engine.plan.execution.nodes.ExecutionNodeExecutor;
+import org.finos.legend.engine.plan.execution.nodes.state.ExecutionState;
+import org.finos.legend.engine.plan.execution.result.ConstantResult;
+import org.finos.legend.engine.plan.execution.result.Result;
+import org.finos.legend.engine.protocol.pure.v1.model.executionPlan.nodes.ExecutionNode;
+import org.finos.legend.engine.shared.core.operational.opentracing.TracedSupplier;
+import org.finos.legend.engine.shared.core.url.StreamProvider;
+import org.finos.legend.engine.shared.core.url.StreamProviderHolder;
+import org.pac4j.core.profile.CommonProfile;
+
+import java.util.List;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.Executors;
+import java.util.concurrent.Semaphore;
+import java.util.concurrent.ThreadPoolExecutor;
+import java.util.concurrent.atomic.AtomicReference;
+
+public final class ConcurrentExecutionNodeExecutorPool
+{
+    private static final AtomicReference<ConcurrentExecutionNodeExecutorPool> INSTANCE = new AtomicReference<>();
+
+    private final int poolSize;
+    private final ThreadPoolExecutor executor;
+    private final Semaphore availableThreads;
```

Comment:
I think we could avoid the semaphore, and set the setRejectedExecutionHandler to `CallerRunsPolicy`, such that if the pool is busy, the caller thread will handle it.  This simplifies too the code around gathering the semaphor and executing locally when cannot acquire.  

Thoughts?

Reply:
This completely makes sense. Though for 1st pass we want design to be on defensive side a bit and not combine execution with parallel pool thread and current thread. Though totally agree this would be a good optimization